filebeat/reader/line/line.go

```diff
@@ -18,177 +18,258 @@
-		logp.Err("Error decoding line: %s", err)
-		// In case of error increase size by unencoded length
-		sz = idx + len(r.nl)
+// msgSize returns the size of the encoded message on the disk
+func (r *decoderReader) msgSize(symlen []int, size int) (int, []int, error) {
+	n := 0
+	for size > 0 {
+		if len(symlen) <= n {
+			return 0, symlen, fmt.Errorf("error calculating size: too short symlen")
+		}
+
+		size -= symlen[n]
+		n++
 	}
 
-	// consume transformed bytes from input buffer
-	err = r.inBuffer.Advance(sz)
-	r.inBuffer.Reset()
+	symlen = symlen[n:]
 
-	// continue scanning input buffer from last position + 1
-	r.inOffset = idx + 1 - sz
-	if r.inOffset < 0 {
-		// fix inOffset if '\n' has encoding > 8bits + firl line has been decoded
-		r.inOffset = 0
-	}
+	return n, symlen, nil
+}
 
-	return err
+func (r *decoderReader) symbolsLen() []int {
+	s := r.symlen
+	r.symlen = []int{}
+	return s
 }
 
-func (r *Reader) decode(end int) (int, error) {
+// conv converts encoded bytes into UTF-8 and produces a symlen array which
+// records the size of the encoded bytes and its converted size
+func (r *decoderReader) conv(in []byte, out []byte) (int, int, error) {
 	var err error
-	buffer := make([]byte, 1024)
-	inBytes := r.inBuffer.Bytes()
-	start := 0
+	nProcessed := 0
+	decodedChar := make([]byte, 64)
+	r.symlen = make([]int, len(in))
 
-	for start < end {
-		var nDst, nSrc int
+	i := 0
+	srcLen := len(in)
+	for i < srcLen {
+		j := i + 1
 
-		nDst, nSrc, err = r.decoder.Transform(buffer, inBytes[start:end], false)
-		if err != nil {
-			// Check if error is different from destination buffer too short
-			if err != transform.ErrShortDst {
-				r.outBuffer.Write(inBytes[0:end])
-				start = end
-				break
+		for j <= srcLen {
+			nDst, nSrc, err := r.decoder.Transform(decodedChar, in[i:j], false)
+			if err != nil {
+				// if no char is decoded, try increasing the input buffer
+				if err == transform.ErrShortSrc {
+					j++
+
+					// if the buffer size cannot be increased, return what's been decoded and an error
+					if srcLen < j {
+						n, _ := r.copyToOut(out)
+						r.symlen = r.symlen[:nProcessed]
+						return n, nProcessed, err
+					}
+				}
+				err = nil
 			}
 
-			// Reset error as decoding continues
-			err = nil
+			// move in the symlen buffer if no char is decoded
+			if nDst == 0 && nSrc == 0 {
+				nProcessed++
+				continue
+			}
+
+			r.symlen[nProcessed] = nDst
+			r.buf.Write(decodedChar[:nDst])
+			nProcessed++
+			break
+		}
+		i = j
+	}
+
+	n, err := r.copyToOut(out)
+	return n, nProcessed, err
+}
+
+func (r *decoderReader) copyToOut(out []byte) (int, error) {
+	until := len(out)
+	if r.buf.Len() < until {
+		until = r.buf.Len()
+	}
+	b, err := r.buf.Collect(until)
+	return copy(out, b), err
+}
+
+type lineScanner struct {
+	in         *decoderReader
+	separator  []byte
+	bufferSize int
+
+	symlen         []int
+	buf            *streambuf.Buffer
+	offset         int
+	bytesOffset    int
+	lastMessageLen int
+}
```

Comment:
types should be at the top of the source file. Given both readers are somewhat non-trivial consider splitting them into separate files.

Reply:
yes, i kind of expected this request, so i put the types like this, so it would be easier to copy them into different files.